[SMALLFIX] Clear MappedByteBuffer explicitly in PacketReaders

peisun1115 opened the pull request.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12817/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12818/
Test PASSed.

apc999 commented:
I didn't see the fundamental difference here. Could you elaborate why this PR can solve OOM (do we have memory leak before)?

peisun1115 commented:
@apc999 DataByteBuffer#release explicitly clears the MemoryMappedBuffer (this is not a netty buffer).   I didn't do this before. We do similar stuff in LocalBlockInStream.java as well.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12821/
Test FAILed.

peisun1115 commented on core/common/src/main/java/alluxio/PropertyKey.java:
```diff
@@ -299,7 +299,7 @@
   USER_UFS_FILE_WRITER_CLASS(Name.USER_UFS_FILE_WRITER_CLASS,
       "alluxio.client.netty.NettyUnderFileSystemFileWriter"),
-  USER_LOCAL_READER_PACKET_SIZE_BYTES(Name.USER_LOCAL_READER_PACKET_SIZE_BYTES, "64KB"),
+  USER_LOCAL_READER_PACKET_SIZE_BYTES(Name.USER_LOCAL_READER_PACKET_SIZE_BYTES, "1MB"),
```
This is increased to achieve similar performance as the legacy one when reading from  local as the impl is similar

peisun1115 commented on core/common/src/main/java/alluxio/network/protocol/databuffer/DataByteBuffer.java:
```diff
@@ -56,6 +56,16 @@ public ByteBuffer getReadOnlyByteBuffer() {
+  @Override
   public void release() {
     if (mBuffer.isDirect()) {
       BufferUtils.cleanDirectBuffer(mBuffer);
```
This is the place that releases the MappedByteBuffer

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12823/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12826/
Test PASSed.

apc999 commented on core/common/src/main/java/alluxio/network/protocol/databuffer/DataBuffer.java:
```diff
@@ -40,6 +40,22 @@
+   * @param dstIndex the first index of the destination
+   * @param length the number of bytes to transfer
+   */
+  void readBytes(byte[] dst, int dstIndex, int length);
```
how do you indicate failure for this method? there is no returned value and no exception thrown in this interface

peisun1115 replied:
It is reading bytes from memory, no checked exception is expected.

apc999 replied:
SG

apc999 commented on core/common/src/main/java/alluxio/network/protocol/databuffer/DataBuffer.java:
```diff
@@ -40,6 +40,22 @@
+  /**
+   * @return the number of readable bytes remaining
+   */
+  int readableBytes();
```
should this be int or long?

peisun1115 replied:
int  since ByteBuffer#limit() and ByteBuf#readableBytes return int.

apc999 replied:
But Java int max is 2,147,483,647 which is 2GB.
ByteBuffer and ByteBuf may have legacy reasons to return int rather than long. But this interface is an abstraction and I don't see why we should be bound by the two current existing subclass implementation? Typically, for length in bytes, we should always use long rather than int.

peisun1115 replied:
Java's array size limit is <= INT_MAX. length should be <= dst.length. So I think int should be used?

peisun1115 replied:
https://docs.oracle.com/javase/specs/jls/se8/html/jls-10.html#jls-10.4

Arrays must be indexed by int values; short, byte, or char values may also be used as index values because they are subjected to unary numeric promotion (§5.6.1) and become int values.

apc999 replied:
SG

apc999 commented on core/common/src/main/java/alluxio/network/protocol/databuffer/DataByteArrayChannel.java:
```diff
@@ -52,6 +52,16 @@ public ByteBuffer getReadOnlyByteBuffer() {
 
   @Override
+  public void readBytes(byte[] dst, int dstIndex, int length) {
+    throw new RuntimeException("DataByteArrayChannel#readBytes is not implemented.");
```
` UnsupportedOperationException` please, same comment for all other `RuntimeException` in this PR.

peisun1115 replied:
done

apc999 commented:
LGTM modulo my comments about exception.

Apart from this PR, I think we should refactor `DataBuffer` into `ReadableDataBuffer` and `WritableDataBuffer`, otherwise this interface is getting overloaded.

peisun1115 commented:
@apc999 I've addressed your comments.  About refactoring, I am not sure whether we should refactor it to readable and writable ones. But I agree, some cleanup for that interface might be necessary

apc999 commented:
left one comment on using long rather than int for the readable bytes length. Otherwise LGTM

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12827/
Test PASSed.

apc999 commented:
LGTM
